openmoltools/forcefield_generators.py

```diff
@@ -519,6 +531,296 @@ def gaffTemplateGenerator(forcefield, residue, structure=None):
+
+        """
+        # Return if empty
+        if not oemols:
+            return
+
+        # Ensure oemols is iterable
+        try:
+            iterator = iter(oemols)
+        except TypeError as te:
+            oemols = [ oemols ]
+
+        # Create copies
+        oemols = [ oemol.CreateCopy() for oemol in oemols ]
+
+        # Cache OEMols
+        from openeye import oechem
+        self._oemols.update( { oechem.OEMolToSmiles(oemol) : oemol for oemol in oemols } )
+
+    @staticmethod
+    def _match_residue(residue, oemol_template):
+        """Determine whether a residue matches an OEMol template and return a list of corresponding atoms.
+
+        This implementation uses NetworkX for graph isomorphism determination.
+
+        Parameters
+        ----------
+        residue : simtk.openmm.app.topology.Residue
+            The residue to check
+        oemol_template : openeye.oechem.OEMol
+            The OEMol template to compare it to
+
+        Returns
+        -------
+        matches : dict of int : int
+            matches[residue_atom_index] is the corresponding OEMol template atom index
+            or None if it does not match the template
+
+        """
+        import networkx as nx
+        from openeye import oechem
+
+        # Make a copy of the template
+        oemol_template = oemol_template.CreateCopy()
+
+        # Ensure atom names are unique
+        oechem.OETriposAtomNames(oemol_template)
+
+        # Build list of external bonds for residue
+        number_of_external_bonds = { atom : 0 for atom in residue.atoms() }
+        for bond in residue.external_bonds():
+            if bond[0] in number_of_external_bonds: number_of_external_bonds[bond[0]] += 1
+            if bond[1] in number_of_external_bonds: number_of_external_bonds[bond[1]] += 1
+
+        # Residue graph
+        residue_graph = nx.Graph()
+        for atom in residue.atoms():
+            residue_graph.add_node(atom, element=atom.element.atomic_number, number_of_external_bonds=number_of_external_bonds[atom])
+        for bond in residue.internal_bonds():
+            residue_graph.add_edge(bond[0], bond[1])
+
+        # Template graph
+        # TODO: We can support templates with "external" bonds or atoms using attached string data in future
+        # See https://docs.eyesopen.com/toolkits/python/oechemtk/OEChemClasses/OEAtomBase.html
+        template_graph = nx.Graph()
+        for oeatom in oemol_template.GetAtoms():
+            template_graph.add_node(oeatom.GetName(), element=oeatom.GetAtomicNum(), number_of_external_bonds=0)
+        for oebond in oemol_template.GetBonds():
+            template_graph.add_edge(oebond.GetBgn().GetName(), oebond.GetEnd().GetName())
+
+        # Determine graph isomorphism
+        from networkx.algorithms import isomorphism
+        graph_matcher = isomorphism.GraphMatcher(residue_graph, template_graph)
+        if graph_matcher.is_isomorphic() == False:
```

Comment:
Not sure if this is intended, but if you want to compare also the nodes' attributes (e.g. `element` and `number_of_external_bonds`), I think you'll have to pass a `node_match` function to `GraphMatcher` because I believe that the default behavior is to ignore them.